Add static type checking with MyPy

cthoyt opened the pull request:
Blocked by #146 

This PR does the following:

- [x] Adds `mypy` environment to `tox.ini`
- [x] Adds relevant configuration to `pyproject.toml` to exclude auto-generated modules
- [x] Adds `mypy` testing to CI configuration
- [x] Fixes all type annotations
- [x]  Fix remaining tests

cthoyt commented:
@hrshdhgd do you want to help me with this one? We should definitely merge #142 first, but then this might be a good experience for you as well

hrshdhgd commented:
On it!

hrshdhgd commented:
I'm assuming you've already taken care of the 1st three bullets?

cthoyt commented:
Yes, the bullets that I checked are already done in this PR

hrshdhgd commented:
I'm facing an interesting dilemma. So `sssom/cli.py:608: error: Need type annotation for "input_file"` . The line is as follows:
```
for input_file in inputs[2:]:
            msdf1 = merged_msdf
```
Now when I add `# type: str` at the end of the for statement flake8 screams at me saying `inline comment should start with '# 'flake8(E262)`

How do I address this?

cthoyt commented on sssom/util.py:
```diff
@@ -316,7 +316,7 @@ def filter_redundant_rows(df: pd.DataFrame, ignore_predicate=False) -> pd.DataFr
     dfmax: pd.DataFrame
     dfmax = df.groupby(key, as_index=False)[CONFIDENCE].apply(max).drop_duplicates()
-    max_conf = {}  # type: Dict[Tuple[Any]]
+    max_conf = {}  # type: Dict[Any, Any]
```
why not `max_conf: Dict[Any, Any] = {}`? I think this syntax was added in py36, which is the minimum version for this ontology

hrshdhgd replied:
Yeah, realized that later and made the change. Thanks for checking!

cthoyt replied:
yes, you're right either are accepted by MyPy, but it's nicer to use the syntax because then it's actually stored for introspection in python in max_conf.__annotations__

hrshdhgd commented:
Some changes that satisfy Mypy results in errors in purest. Is that expected?

hrshdhgd commented on sssom/util.py:
```diff
@@ -885,13 +885,14 @@ def to_mapping_set_dataframe(doc: MappingSetDocument) -> MappingSetDataFrame:
-            if mdict[key]:
-                m[key] = mdict[key]
-        data.append(m)
+    if doc.mapping_set.mappings is not None:
```
@cthoyt , does this makes sense. This is the fix for the error:

`Item "None" of "Union[Dict[Any, Any], Mapping, List[Union[Dict[Any, Any], Mapping]], None]" has no attribute "__iter__" (not iterable)`

cthoyt replied:
we should really fix #145 before doing more on here otherwise we're going to get merge problems. I'll comment on this one after we finish 145

cthoyt commented:
> Some changes that satisfy Mypy result in errors in pytest. Is that expected?

Not sure what you mean - let me know if you see this being reproduced in the CI tests. For the most part, changing type annotations doesn't change executed code (except in the case of forward references... you can google that if you do end up with an issue like this)

hrshdhgd commented:
> > Some changes that satisfy Mypy result in errors in pytest. Is that expected?
> 
> Not sure what you mean - let me know if you see this being reproduced in the CI tests. For the most part, changing type annotations doesn't change executed code (except in the case of forward references... you can google that if you do end up with an issue like this)

`sssom/util.py:67: error: Incompatible types in assignment (expression has type "None", variable has type "Dict[str, Any]")`

So I fix it by changing 
`    prefixmap: Dict[str, Any] = None  # maps CURIE prefixes to URI bases` 
to 
`prefixmap: Dict[str, Any] = {}  # maps CURIE prefixes to URI bases`

fixes Mypy but when I run purest, this fails.

```
____________________________________________________________________ ERROR collecting tests/test_cli.py _____________________________________________________________________
tests/test_cli.py:6: in <module>
    from sssom.cli import (
sssom/__init__.py:3: in <module>
    from .util import (  # noqa:401
sssom/util.py:61: in <module>
    class MappingSetDataFrame:
/opt/anaconda3/envs/sssom-new-linkml/lib/python3.8/dataclasses.py:1019: in dataclass
    return wrap(cls)
/opt/anaconda3/envs/sssom-new-linkml/lib/python3.8/dataclasses.py:1011: in wrap
    return _process_class(cls, init, repr, eq, order, unsafe_hash, frozen)
/opt/anaconda3/envs/sssom-new-linkml/lib/python3.8/dataclasses.py:861: in _process_class
    cls_fields = [_get_field(cls, name, type)
/opt/anaconda3/envs/sssom-new-linkml/lib/python3.8/dataclasses.py:861: in <listcomp>
    cls_fields = [_get_field(cls, name, type)
/opt/anaconda3/envs/sssom-new-linkml/lib/python3.8/dataclasses.py:745: in _get_field
    raise ValueError(f'mutable default {type(f.default)} for field '
E   ValueError: mutable default <class 'dict'> for field prefixmap is not allowed: use default_factory
```

hrshdhgd commented:
> fixes Mypy but when I run pytest, this fails.

The fix seems to be `prefixmap: Dict[str, Any] = field(default_factory=dict)  # maps CURIE prefixes to URI bases` but I have never seen something like this before. Thoughts?

cthoyt commented:
> > fixes Mypy but when I run pytest, this fails.
> 
> The fix seems to be `prefixmap: Dict[str, Any] = field(default_factory=dict) # maps CURIE prefixes to URI bases` but I have never seen something like this before. Thoughts?
> 
> [Source](https://stackoverflow.com/questions/53632152/why-cant-dataclasses-have-mutable-defaults-in-their-class-attributes-declaratio)

Yes that's the fix! Welcome to the wild world of dataclasses

cthoyt commented:
@hrshdhgd please stop making updates here until you review #145. This will make more work to merge after if you keep going

cthoyt commented:
@hrshdhgd thanks for being patient - I merged the other PR then updated into this one, so we can continue here

hrshdhgd commented:
I've made as many changes as I could based on my knowledge of the project. there still are 51 errors where I'll need help with data type declaration decisions. Most of them are redundant and fixing one will fix multiple of them.

```
sssom/sparql_util.py:82: error: Argument "prefixmap" to "MappingSetDataFrame" has incompatible type "Optional[Dict[str, str]]"; expected "Dict[str, str]"
sssom/parsers.py:42: error: Argument 1 to "read_pandas" has incompatible type "str"; expected "TextIO"
sssom/parsers.py:63: error: Argument "curie_map" to "_get_curie_map_and_metadata" has incompatible type "Optional[Dict[str, Any]]"; expected "Dict[str, Any]"
sssom/parsers.py:63: error: Argument "meta" to "_get_curie_map_and_metadata" has incompatible type "Optional[Dict[str, Any]]"; expected "Dict[str, Any]"
sssom/parsers.py:64: error: Argument "curie_map" to "_get_curie_map_and_metadata" has incompatible type "Optional[Dict[str, Any]]"; expected "Dict[str, Any]"
sssom/parsers.py:64: error: Argument "meta" to "_get_curie_map_and_metadata" has incompatible type "Optional[Dict[str, Any]]"; expected "Dict[str, Any]"
sssom/parsers.py:66: error: Argument "curie_map" to "from_sssom_dataframe" has incompatible type "Optional[Dict[str, Any]]"; expected "Dict[str, str]"
sssom/parsers.py:66: error: Argument "meta" to "from_sssom_dataframe" has incompatible type "Optional[Dict[str, Any]]"; expected "Dict[str, str]"
sssom/parsers.py:80: error: Argument "curie_map" to "_get_curie_map_and_metadata" has incompatible type "Optional[Dict[str, str]]"; expected "Dict[str, Any]"
sssom/parsers.py:80: error: Argument "meta" to "_get_curie_map_and_metadata" has incompatible type "Optional[Dict[str, str]]"; expected "Dict[str, Any]"
sssom/parsers.py:98: error: Argument "curie_map" to "_get_curie_map_and_metadata" has incompatible type "Optional[Dict[str, str]]"; expected "Dict[str, Any]"
sssom/parsers.py:98: error: Argument "meta" to "_get_curie_map_and_metadata" has incompatible type "Optional[Dict[str, str]]"; expected "Dict[str, Any]"
sssom/parsers.py:102: error: Argument "curie_map" to "from_sssom_json" has incompatible type "Optional[Dict[str, str]]"; expected "Dict[str, str]"
sssom/parsers.py:121: error: Argument "curie_map" to "_get_curie_map_and_metadata" has incompatible type "Optional[Dict[str, str]]"; expected "Dict[str, Any]"
sssom/parsers.py:121: error: Argument "meta" to "_get_curie_map_and_metadata" has incompatible type "Optional[Dict[str, str]]"; expected "Dict[str, Any]"
sssom/parsers.py:126: error: Argument 2 to "from_obographs" has incompatible type "Optional[Dict[str, str]]"; expected "Dict[str, str]"
sssom/parsers.py:136: error: Incompatible types in assignment (expression has type "Mapping[str, Any]", variable has type "Dict[str, Any]")
sssom/parsers.py:139: error: Incompatible types in assignment (expression has type "Mapping[str, Any]", variable has type "Dict[str, Any]")
sssom/parsers.py:212: error: Incompatible types in assignment (expression has type "List[Mapping]", variable has type "Union[Dict[Any, Any], Mapping, List[Union[Dict[Any, Any], Mapping]], None]")
sssom/parsers.py:250: error: Argument 2 to "curie_from_uri" has incompatible type "Optional[Dict[str, str]]"; expected "Mapping[str, str]"
sssom/parsers.py:263: error: Argument 2 to "curie_from_uri" has incompatible type "Optional[Dict[str, str]]"; expected "Mapping[str, str]"
sssom/parsers.py:287: error: Incompatible types in assignment (expression has type "List[Mapping]", variable has type "Union[Dict[Any, Any], Mapping, List[Union[Dict[Any, Any], Mapping]], None]")
sssom/parsers.py:288: error: Argument "metadata" to "_set_metadata_in_mapping_set" has incompatible type "Optional[Dict[str, str]]"; expected "Dict[str, str]"
sssom/parsers.py:289: error: Argument "curie_map" to "MappingSetDocument" has incompatible type "Optional[Dict[str, str]]"; expected "Dict[str, str]"
sssom/parsers.py:301: error: Argument "metadata" to "_set_metadata_in_mapping_set" has incompatible type "Optional[Dict[str, str]]"; expected "Dict[str, str]"
sssom/parsers.py:354: error: Incompatible types in assignment (expression has type "List[Mapping]", variable has type "Union[Dict[Any, Any], Mapping, List[Union[Dict[Any, Any], Mapping]], None]")
sssom/parsers.py:439: error: Incompatible types in assignment (expression has type "List[Mapping]", variable has type "Union[Dict[Any, Any], Mapping, List[Union[Dict[Any, Any], Mapping]], None]")
sssom/parsers.py:440: error: Argument "metadata" to "_set_metadata_in_mapping_set" has incompatible type "Optional[Dict[str, str]]"; expected "Dict[str, str]"
sssom/parsers.py:617: error: Incompatible types in assignment (expression has type "List[Mapping]", variable has type "Union[Dict[Any, Any], Mapping, List[Union[Dict[Any, Any], Mapping]], None]")
sssom/parsers.py:674: error: Argument "meta" to "from_sssom_dataframe" has incompatible type "Optional[Dict[str, str]]"; expected "Dict[str, str]"
sssom/writers.py:60: error: Item "None" of "Optional[Any]" has no attribute "to_csv"
sssom/writers.py:131: error: Item "None" of "Union[Dict[Any, Any], Mapping, List[Union[Dict[Any, Any], Mapping]], None]" has no attribute "__iter__" (not iterable)
sssom/rdf_util.py:22: error: Need type annotation for "rewire_map" (hint: "rewire_map: Dict[<type>, <type>] = ...")
sssom/rdf_util.py:28: error: Item "None" of "Union[Dict[Any, Any], Mapping, List[Union[Dict[Any, Any], Mapping]], None]" has no attribute "__iter__" (not iterable)
sssom/rdf_util.py:29: error: Item "Dict[Any, Any]" of "Union[Any, Dict[Any, Any], Mapping]" has no attribute "predicate_id"
sssom/rdf_util.py:31: error: Item "Dict[Any, Any]" of "Union[Any, Dict[Any, Any], Mapping]" has no attribute "subject_id"
sssom/rdf_util.py:31: error: Item "Dict[Any, Any]" of "Union[Any, Dict[Any, Any], Mapping]" has no attribute "object_id"
sssom/rdf_util.py:33: error: Item "Dict[Any, Any]" of "Union[Any, Dict[Any, Any], Mapping]" has no attribute "object_id"
sssom/rdf_util.py:33: error: Item "Dict[Any, Any]" of "Union[Any, Dict[Any, Any], Mapping]" has no attribute "subject_id"
sssom/rdf_util.py:39: error: Item "None" of "Union[Any, str, EntityId, None]" has no attribute "split"
sssom/rdf_util.py:50: error: Argument 1 to "expand_curie" has incompatible type "Union[Any, str, EntityId, None]"; expected "str"
sssom/cliques.py:23: error: Item "Dict[Any, Any]" of "Union[Any, Dict[Any, Any], Mapping]" has no attribute "subject_id"
sssom/cliques.py:24: error: Item "Dict[Any, Any]" of "Union[Any, Dict[Any, Any], Mapping]" has no attribute "object_id"
sssom/cliques.py:25: error: Item "Dict[Any, Any]" of "Union[Any, Dict[Any, Any], Mapping]" has no attribute "predicate_id"
sssom/cliques.py:78: error: Item "None" of "Union[Dict[Any, Any], Mapping, List[Union[Dict[Any, Any], Mapping]], None]" has no attribute "__iter__" (not iterable)
sssom/cliques.py:79: error: Item "Dict[Any, Any]" of "Union[Any, Dict[Any, Any], Mapping]" has no attribute "subject_id"
sssom/cliques.py:81: error: Item "Dict[Any, Any]" of "Union[Dict[Any, Any], Mapping, List[Union[Dict[Any, Any], Mapping]], None]" has no attribute "append"
sssom/cliques.py:81: error: Item "None" of "Union[Dict[Any, Any], Mapping, List[Union[Dict[Any, Any], Mapping]], None]" has no attribute "append"
sssom/io.py:27: error: Unexpected keyword argument "serialisation"
sssom/io.py:52: error: Argument "curie_map_mode" to "get_metadata_and_curie_map" has incompatible type "Optional[str]"; expected "str"
sssom/cli.py:334: error: Missing positional arguments "url", "graph", "predmap", "predicates", "limit", "curie_map" in call to "EndpointConfig"
Found 51 errors in 7 files (checked 17 source files)
```

cthoyt commented on sssom/cli.py:
```diff
@@ -573,7 +575,7 @@ def merge(inputs: Tuple[str, str], output: TextIO, reconcile: bool = True):
     # and merge them into the merged file above
     if len(inputs) > 2:
-        for input_file in inputs[2:]:
+        for input_file in inputs[2:]:  # type: str
```
I don't like this, what is it supposed to refer to? let's strictly only use python syntax for type annotations, and never the comment syntax

hrshdhgd replied:
That's what I thought too. But it was complaining about the type and hence I made that comment to address it.

cthoyt replied:
Oh I see now. the type annotation is wrong for inputs. it should be `Sequence[str]`, not `Tuple[str, str]`

hrshdhgd replied:
I made the change and committing it. Thank you so much @cthoyt !

cthoyt replied:
Thanks :)

cthoyt commented on sssom/cliques.py:
```diff
@@ -18,41 +18,42 @@ def to_networkx(msdf: MappingSetDataFrame) -> nx.DiGraph:
+            p = mapping.predicate_id
+            # TODO: this is copypastad from export_ptable
+
+            pi = None
```
what happened to all of that hard work on improving the variable names here?

hrshdhgd replied:
Any idea @matentzn ? I remember the refactoring of all these in a different branch. I thought merging down from `main` would have reflected those changes. Also I just worked off of the branch you had @cthoyt .

cthoyt replied:
you can see it's because the code was copy-pasted, and I only fixed it in one place. leave it for now

matentzn replied:
Maybe cliques.py got a chris PR in the meantime. Unless PR howls at bad variable names, we should expect those coming back from time to time.

cmungall replied:
yeh I agree this is horrible... but I wouldn't block this PR on it. We should take my code factor it into the logic (mapping mappings to logical axioms) and then the ptable-format specific stuff. I will be better in future.

cthoyt replied:
Agreed we will fix this in a future refactoring PR

cthoyt replied:
See #152

cthoyt commented on sssom/cliques.py:
```diff
@@ -18,41 +18,42 @@ def to_networkx(msdf: MappingSetDataFrame) -> nx.DiGraph:
+            s = mapping.subject_id
+            o = mapping.object_id
+            p = mapping.predicate_id
+            # TODO: this is copypastad from export_ptable
```
TODO indeed. please refactor the code from export_ptable and reuse it

hrshdhgd replied:
@matentzn ?

cthoyt replied:
This is TODO for another PR, sorry for the confusion

cthoyt commented on sssom/cliques.py:
```diff
@@ -82,8 +82,8 @@ def split_into_cliques(msdf: MappingSetDataFrame):
-def invert_dict(d: dict) -> dict:
-    invdict = {}
+def invert_dict(d: Dict[str, str]) -> Dict[str, str]:
+    invdict = {}  # type: Dict[Any, Any]
```
use syntax-based type annotation

```suggestion
     invdict: Dict[Any, Any] = {}
```

We should also be able to do better than Any, Any

hrshdhgd replied:
maybe Dict[str, str] ? But I do not know the general data format of objects expected for this function. Thoughts @matentzn ?

cthoyt replied:
@hrshdhgd not knowing is just the first step on the journey to knowing! The way I figure this stuff out is to look through the codebase and see where the function is used (you can basically act like a type checker yourself) and see what goes in then what comes out.

matentzn replied:
No ideas about the clique.py code. Ask Chris, but only if these are PR breakers.

hrshdhgd replied:
Figured this out!

cthoyt commented on sssom/parsers.py:
```diff
@@ -33,7 +33,7 @@
 
 def read_sssom_table(
-    file_path: str, curie_map: Dict[str, str] = None, meta: Dict[str, str] = None
+    file_path: str, curie_map: Dict[str, Any] = None, meta: Dict[str, Any] = None
```
try to do better than Any

cthoyt commented on sssom/parsers.py:
```diff
@@ -590,41 +591,44 @@ def to_mapping_set_document(msdf: MappingSetDataFrame) -> MappingSetDocument:
-        mlist.append(m)
+    if msdf.df is not None:
+        for _, row in msdf.df.iterrows():
+            mdict = {}
```
the code in the innermost loop should be a class method on `Mapping`

matentzn replied:
Issue

cthoyt commented on sssom/sparql_util.py:
```diff
@@ -27,18 +27,18 @@ def query_mappings(config: EndpointConfig) -> MappingSetDataFrame:
     sparql = SPARQLWrapper(config.url)
     if config.graph is None:
-        g = "?g"
+        g = "?g"  # type:URIRef
```
this doesn't make sense

cthoyt commented on sssom/util.py:
```diff
@@ -169,6 +176,8 @@ def load(self, filename) -> None:
         self.df = read_pandas(filename)
 
     def convert(self) -> Dict[str, Any]:
+        if self.df is None:
```
why is the df even allowed to be none?

matentzn replied:
I would prefer not to add to the complexity of this PR.. This is a separate issue

cthoyt commented on sssom/util.py:
```diff
@@ -308,7 +317,7 @@ def filter_redundant_rows(df: pd.DataFrame, ignore_predicate=False) -> pd.DataFr
     dfmax: pd.DataFrame
     dfmax = df.groupby(key, as_index=False)[CONFIDENCE].apply(max).drop_duplicates()
-    max_conf = {}
+    max_conf: Dict[Any, Any] = {}
```
better annotation here

hrshdhgd replied:
maybe Dict[str, str] ? But I do not know the general data format of objects expected for this function. Thoughts @matentzn ?

matentzn replied:
No thoughts, sorry.. You will have to check what is expected by the functions its going to.

hrshdhgd replied:
Done!

cthoyt commented on sssom/writers.py:
```diff
@@ -49,7 +49,7 @@ def write_table(msdf: MappingSetDataFrame, file: TextIO, serialisation="tsv") ->
 
     if msdf.metadata is not None:
-        meta = {k: v for k, v in msdf.metadata.items()}
+        meta = {k: v for k, v in msdf.metadata.items()}  # type: Dict[Any, Any]
```
better type annotation, use syntax

matentzn commented:
Hey both, this PR lets my heart race in happiness - please let me know if you need me to take over and finish it - we should merge asap and make a new release, and hopefully than this was the last of the massive refactor PRs for a while :) thank you both.

Just one thing: lets please strictly adhere to mypy here and not address any other issues. If you find other stuff, make issue for them.

cthoyt commented:
@matentzn @hrshdhgd I've finished updating all of the type checking (only in a few places using ignores, such as the crazy typing in the autogenerated code). There were a lot of places where the handling of falsy prefix maps (e.g., None or empty) were not explicit. This made it really hard to figure out what the business logic should be. Because I've done what it took to make the program make sense, there are now a few failing tests, so I'd suggest Nico goes through the deep set of functions handling these prefix maps and figures out which business logic should be the right one.

I know it was your goal to do minimum code changes, but type checking exposes logical errors that fully call for refactoring and bug fixing, so we're just lucky it wasn't more

cthoyt commented on tests/test_parsers.py:
```diff
@@ -52,7 +52,7 @@ def setUp(self) -> None:
         self.alignmentxml_file = f"{test_data_dir}/oaei-ordo-hp.rdf"
         self.alignmentxml = minidom.parse(self.alignmentxml_file)
-        self.metadata, self.curie_map = get_default_metadata()
+        self.curie_map, self.metadata = get_default_metadata()
```
@hrshdhgd good catch - these were mixed up in so many places that I ended up replacing it with a named tuple to try and standardize it

matentzn replied:
Wouldnt this be wrong here then? Doesnt get_default_metadata() return a `Metadata` object?

cthoyt replied:
A Metadata object is a namedtuple, therefore you can directly unpack it if you want

cthoyt commented on sssom/util.py:
```diff
@@ -330,7 +330,7 @@ def filter_redundant_rows(
     dfmax: pd.DataFrame
     dfmax = df.groupby(key, as_index=False)[CONFIDENCE].apply(max).drop_duplicates()
-    max_conf: Dict[Any, Any] = {}
+    max_conf: Dict[Tuple, float] = {}
```
tuple of what?

hrshdhgd replied:
Tuple[str, ...], fixed!

cthoyt commented:
Everyone can take a look tomorrow to see if any of the changes we made weren't covered by tests. Otherwise, let's get this merged and move on with more fun refactoring in a future PR!

matentzn reviewed (approved):
This is a huge applause to both of you. I am impressed - some of the python is a good deal beyond my current ability and just reading this PR has thought me a great deal. 

If we could make sure that `make test` also runs the mypy, that'd be great.

cthoyt commented:
> If we could make sure that `make test` also runs the mypy, that'd be great.

In 778bfca, I added mypy to the default list of environments that's run with tox. Since the `tox` command is run in the `test` goal, this does the trick! I also added a mypy-specific goal in f75b724
